microprofile-open-api/src/test/java/org/jboss/eap/qe/microprofile/openapi/advanced/AcceptHeadersAndFormatParameterInteractionTest.java

```diff
@@ -0,0 +1,365 @@
+package org.jboss.eap.qe.microprofile.openapi.advanced;
+
+import static io.restassured.RestAssured.get;
+import static io.restassured.RestAssured.given;
+
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import javax.ws.rs.core.MediaType;
+
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.arquillian.api.ServerSetup;
+import org.jboss.as.arquillian.api.ServerSetupTask;
+import org.jboss.as.arquillian.container.ManagementClient;
+import org.jboss.eap.qe.microprofile.openapi.OpenApiServerConfiguration;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.ProviderApplication;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.RoutingServiceConstants;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.api.DistrictService;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.data.DistrictEntity;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.model.District;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.rest.DistrictsResource;
+import org.jboss.eap.qe.microprofile.openapi.apps.routing.provider.services.InMemoryDistrictService;
+import org.jboss.eap.qe.microprofile.openapi.v10.OpenApi10OnJaxRsAnnotationsTest;
+import org.jboss.eap.qe.microprofile.tooling.server.configuration.creaper.ManagementClientProvider;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.wildfly.extras.creaper.core.online.OnlineManagementClient;
+
+/**
+ * Test cases to verify negative scenario in which "Accept" HTTP header and {@code format} query parameter are
+ * conflicting or not valid.
+ * {@link MediaType} constants are used as HTTP "Accept" header values.
+ */
+@RunWith(Arquillian.class)
+@ServerSetup({ AcceptHeadersAndFormatParameterInteractionTest.OpenApiExtensionSetup.class })
+@RunAsClient
```

Comment:
As all tests are `@RunAsClient` and test itself is not deployed into server with deployment there is no need to use `@ServerSetup`. Could you modify the test so setup() and tearDown() are annotated by standard JUnit `@BeforeClass` and `@AfterClass` annotation (and moved from inner class), please?

Reply:
Well, actually that usage of `ServerSetupTask` is IMO _very_ necessary, at least for MP OpenAPI tests, because it is used to properly control the _server_ configuration with respect to the managed deployment life cycle, so it is not a discriminant the fact that the test is not deployed at server side, see JavaDoc at [1].
I discussed this on the working group chat a couple of weeks ago, here's an excerpt for you to understand the details of what I ran into:

```
...
My issue:
* I was having two warnings from Weld because it could not link two registered MP OpenAPI extension classes implementing OASModelReader and OASFilter interfaces for programmatic OpenAPI doc manipulation 
* first WARN pops up when Arquillian archive is deployed => possible root cause: @BeforeClass happens later than deploy and I enable openapi extension/subsystem in @BeforeClass, so: a deployment contains and registers concrete Filter/ModelReder, Weld tries and load them but the extension is still disabled. If such reasoning is true, enabling the extension should happen soon before deployment.
* second (and last) warn pops up when calling CLI reload command. This happens in @AfterClass, soon after disabling the extension. This means that Weld processes again the - still undeployed deployment containing registered implementations for Filter and ModelReader and emits the same warning. This means that disabling the extension should happen soon after undeployment.
...
```

Similar Arquillian deployment life cycle management issues were observed in some different test cases (for one of those I filed a WIP issue which was solved with this configuration improvement) and were all related to subsystem configuration code being executed in `@BeforeClass` and `@AfterClass`.

Bottom line, enabling and/or disabling the subsystem is - at least for MP OpenAPI spec - a server configuration task that must be synchronous with related server extension life cycle management activities that could affect the deployment phase (e.g.: usage of registered interface implementation as in the above mentioned use case).

I have discussed the thing with developer, which helped in analyzing what was going on and I made some research in Wildfly TS too, where `ServerSetupTask` is used to configure server even when tests are to be run by the client. 

Finally, once that all of the above mentioned rationale was validated, I applied the configuration to all of the MP OpenAPI tests in order to follow the same pattern throughout the module. 

So, to be honest I don't see any value for changing this logical pattern with JUint `@BeforeClass` and `@AfterCless` methods and indeed I believe this is actually unwanted.

Let me know your thoughts.

[1]
https://github.com/wildfly/wildfly-arquillian/blob/master/common/src/main/java/org/jboss/as/arquillian/api/ServerSetupTask.java

Reply:
@fabiobrz Thanks for explanation, now I can see the issue and the motivation for it. So when deployment is managed by Arquillian then it's deployed before `@BeforeClass` is called (so before OpenApi is enabled). 

However the issue can be workaround by manual deployment like:
```
@ArquillianResource
private Deployer deployer;

@Deployment(name = FIRST_DEPLOYMENT, managed = false)
...

 @BeforeClass
public static void setup() throws ManagementClientRelatedException {
        onlineManagementClient = ManagementClientProvider.onlineStandalone();
        OpenApiServerConfiguration.enableOpenApi(onlineManagementClient);
        deployer.deploy(FIRST_DEPLOYMENT);
}
```

I'm unhappy that `managementClient` is passed as parameter into setup(...) method and then it's not used and instead there is used Creaper `onlineManagementClient`:
```
static class OpenApiExtensionSetup implements ServerSetupTask {

        @Override
        public void setup(ManagementClient managementClient, String containerId) throws Exception {
            //  MP OpenAPI up
            onlineManagementClient = ManagementClientProvider.onlineStandalone();
            OpenApiServerConfiguration.enableOpenApi(onlineManagementClient);
        }
```
the code looks weird. 

@rsvoboda @istraka wdyt?

Reply:
Hi @mnovak1 - thanks for your reply, see my comments here.

> However the issue can be workaround by manual deployment like:
...

Well yes, it possibly could. But I am thinking why it should, as `ServerSetupTask` implementation is the tool that Arquillian is providing to the client in order to manage server configuration task _within_ the logical scope of it's (Arquillian's) managed deployment process. So I am trying and find any reason for changing it all again in favor of a more generic - i.e. JUnit - implementation. But feel free to let me know any rationale for it.   

> I'm unhappy that managementClient is passed as parameter into setup(...) method and then it's not used and instead there is used Creaper onlineManagementClient:

So you're suggesting not to use our tooling for server configuration here because a `ManagementClient` instance is already injected? I am actually not sure that the API will prove to be sufficient for our current requirements - as our tooling has proved to be, but I have actually not tried. WDYT?

Reply:
After discussion in chat, current implementation is now using the following wrapping logic provided by new method in `ManagementClientProvider`:

```
...
org.wildfly.extras.creaper.core.ManagementClient.onlineLazy(
                OnlineOptions.standalone().wrap(managementClient.getControllerClient()))
...
```

So `ServerSetupTask#setup()` and `ServerSetupTask#tearDown()` methods implementation can use the `ManagementClient` instance passed as input parameter in order to get a Creaper `OnlineManagmentClient` out of Arquillian `ManagementClient` instance.
This is more direct and clean, although eventually relying on the same configuration source - i.e. `arquillian.xml`.

Due to the following behavior, explained by the `OnlineOptions.standalone().wrap()` Javadoc:
```
         * <p>The {@link org.wildfly.extras.creaper.core.online.OnlineManagementClient} that will be created will wrap
         * the provided {@link org.jboss.as.controller.client.ModelControllerClient modelControllerClient}.</p>
         *
         * <p><b>Note that doing this ties together the lifecycles of the {@code OnlineManagementClient} and
         * the wrapped {@code ModelControllerClient}!</b> If you close the {@code ModelControllerClient}, the
         * {@code OnlineManagementClient} becomes unusable. If you close the {@code OnlineManagementClient},
         * the wrapped {@code ModelControllerClient} will be closed as well.</p>
         *
         * <p>This also means that the new {@code OnlineManagementClient} <b>can't be {@code reconnect}ed</b>, which
         * is sometimes needed if server reload or restart is performed. Hence, use client wrapping with caution!</p>
```
the `OnlineManagmentClient` instance obtained as a wrapper for `ModelControllerClient` instance is never closed, either implicitly or explicitly: e.g.:

```
static class OpenApiExtensionSetup implements ServerSetupTask {

        @Override
        public void setup(ManagementClient managementClient, String containerId) throws Exception {
            OpenApiServerConfiguration.enableOpenApi(ManagementClientProvider.onlineStandalone(managementClient));
        }

        @Override
        public void tearDown(ManagementClient managementClient, String containerId) throws Exception {
            OpenApiServerConfiguration.disableOpenApi(ManagementClientProvider.onlineStandalone(managementClient));
        }
    }
```
as this would cause the underlying `ModelControllerClient` instance too, which would affect Arquillian configuration globally.

Reply:
@fabiobrz thanks Fabio, I don't think closing is left on GC but on `@ServerSetupTask` but this would need closer look. The main point here that it must be closed manually in the test which is common thing when using `@ServerSetupTask`